janus_server/src/datastore.rs

```diff
@@ -0,0 +1,318 @@
+use prio::codec::{decode_u16_items, encode_u16_items, CodecError, Decode};
+use tokio_postgres::{error::SqlState, IsolationLevel, Row};
+
+use crate::message::{Extension, HpkeCiphertext, Nonce, Report, TaskId, Time};
+use std::{future::Future, io::Cursor, pin::Pin};
+use tokio::sync::Mutex;
+
+// TODO(brandon): connection pooling (currently all concurrent transactions in a single Datastore are serialized)
+// TODO(brandon): prepare DB statements
+
+/// Datastore represents a datastore for Janus, with support for transactional reads and writes.
+/// In practice, Datastore instances are currently backed by a PostgreSQL database.
+pub struct Datastore {
+    db: Mutex<tokio_postgres::Client>,
+}
+
+impl Datastore {
+    /// new creates a new Datastore using the given Client for backing storage. It is assumed that
+    /// the Client is connected to a database with a compatible version of the Janus database schema.
```

Comment:
We should figure out DB schema versioning so that we can validate at startup that the DB we connect to has the expected schema. Of course that would probably be a feature of the DB migration tool we adopt so perhaps we shouldn't solve that ourselves.

Reply:
That's a good idea, but we do need to know how we're going to version database schemas first.

We'd also need to be careful to permit live DB schema upgrades -- if we're too restrictive, upgrading to version N+1 would cause running processes expecting version N very unhappy.

The best solution might be to set expectations that all DB schema migrations are done in a way that supports no-downtime migration (i.e. we arrange releases so that we can live-migrate to the new schema while running jobs can work with the DB), but especially in early development we're likely to want to make non-backwards-compatible changes to the schema and doing the release dances to allow these live is probably too much work.